### ISSUE TITLE
langchain dumps not always implemented - Improve langchain runnable serialization

### ISSUE THREAD

**Issue by yairyairyair** [langchain-ai/langchain#25664]
### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the LangChain documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangChain rather than my code.
- [X] The bug is not resolved by updating to the latest stable version of LangChain (or the specific integration package).

### Example Code

```python
from langchain_openai import ChatOpenAI
from langchain_core.prompts import PromptTemplate
from langchain.agents import AgentExecutor, create_react_agent
from langchain.tools import BaseTool
from langchain_core.load import dumps

prompt = PromptTemplate.from_template("Answer the following questions as best you can. You have access to the following tools:\n\n{tools}\n\nUse the following format:\n\nQuestion: the input question you must answer\nThought: you should always think about what to do\nAction: the action to take, should be one of [{tool_names}]\nAction Input: the input to the action\nObservation: the result of the action\n... (this Thought/Action/Action Input/Observation can repeat N times)\nThought: I now know the final answer\nFinal Answer: the final answer to the original input question\n\nBegin!\n\nQuestion: {input}\nThought:{agent_scratchpad}")

# Choose the LLM to use
llm = ChatOpenAI(model="gpt-3.5-turbo")

tools: list[BaseTool] = []

# Construct the ReAct agent
agent = create_react_agent(llm, tools, prompt)

# dumps to a file
with open('react-ai-agent.json', 'w') as f:
    f.write(dumps(agent))


```

### Error Message and Stack Trace (if applicable)

```python
RUNNABLE_FILE_PATH = "./react-ai-agent.json"
with open(RUNNABLE_FILE_PATH, 'r') as f:
  agent = loads(f.read(), secrets_from_env=True)
```

```
NotImplementedError: Trying to load an object that doesn't implement serialization: {'lc': 1, 'type': 'not_implemented', 'id': ['langchain_core', 'runnables', 'base', 'RunnableLambda'], 'repr': "RunnableLambda(lambda x: format_log_to_str(x['intermediate_steps']))"}
```

### Description

Basically i just want to dumps a runnable and loads it, i want it to work for any langchain runnable

### System Info

generic langchain installation

**Comment by yairyairyair** [comment]
If to be more specific these are the not_implemented parts in the json
```json
              "agent_scratchpad": {
                "lc": 1,
                "type": "not_implemented",
                "id": ["langchain_core", "runnables", "base", "RunnableLambda"],
                "repr": "RunnableLambda(lambda x: format_log_to_str(x['intermediate_steps']))"
              }
```

```json
"last": {
      "lc": 1,
      "type": "not_implemented",
      "id": [
        "langchain",
        "agents",
        "output_parsers",
        "react_single_input",
        "ReActSingleInputOutputParser"
      ],
      "repr": "ReActSingleInputOutputParser()",
      "name": "ReActSingleInputOutputParser",
...
}
```

**Comment by baskaryan** [comment]
@yairyairyair thanks for the feedback.

Runnables can contain arbitrary logic (you can wrap any python function in a Runnable), so the only way to be able to serialize *anything* would be to do something unsafe like pickling python objects. 

Out of curiosity, what are you using serialization for?

**Comment by yairyairyair** [comment]
Tried using pickle, dill and jsonpickle all didnt work.
Dont mind the security aspect of it.
Using it to move runnables from one service to another.

**Comment by asheeshgarg** [comment]
Any update on this? This is blocker for me as well.

**Comment by robos4n** [comment]
I've also got a use case where we'd like to persist the agent / associated tools as json strings and then load the agents via json.

**Comment by RogerAylagas** [comment]
This issue is still a blocker for our project. We would appreciate a lot an update on this and an implementation that enables us to serialize our chains. Thank you so much.
